### PR TITLE
Add support for PEBS and MTL

### PR DESCRIPTION
**Pull request by weilinwa**
This PR add support to the perf converter script for MTL platform. It also include a quick solution for $PEBS term that borrowed from our csv parsing script.

### CODE REVIEW

File: scripts/create_perf_json.py [hunk 1]
```diff
@@ -632,6 +640,28 @@ def smi_json(self) -> metric.MetricGroup:
...
+                pebs = m.group(3)
+                alternative = m.group(4).strip()
+
+                new_string = f' {min_max}( {main_event}*{pebs} , {main_event} * {alternative} ) '
+                new_formula = re.sub(m.re, new_string, new_formula)
+
+        formula_list = re.split(OPS, new_formula)
+        for element in formula_list:
+            element = element.strip()
+            if '$PEBS' in element:
```
Comment (captain5050) [inline scripts/create_perf_json.py:660]: The only non-min/max case is Lock_Latency. I think also that max is never used, just min.

File: scripts/create_perf_json.py [hunk 2]
```diff
@@ -1510,6 +1552,9 @@ def append_to_desc(s: str):
                 for em in json.load(extra_json):
                     if em['MetricName'] in skip:
                         continue
+                    if em['MetricName'] in ignore:
+                        _verboseprint2(f"Skipping {em['MetricName']}")
+                        continue
```
Comment (captain5050) [inline scripts/create_perf_json.py:1557]: Perhaps this should be a separate commit

File: scripts/create_perf_json.py [hunk 3]
```diff
@@ -1201,12 +1238,14 @@ def save_form(name, group, form, desc, locate, scale_unit, threshold,
...
+                             'num_dies', 'num_cpus_online', 'PEBS']:
                         continue
                     if v.startswith('tma_') or v.startswith('topdown\\-'):
                         continue
-                    assert v in events or v.upper() in events or v in infoname or v in aux, \
-                        f'Expected {v} to be an event in "{name}": "{form}" on {self.shortname}'
+                    #assert v in events or v.upper() in events or v in infoname or v in aux, \
+                    #    f'Expected {v} to be an event in "{name}": "{form}" on {self.shortname}'
+                    if v not in events and v.upper() not in events and v not in infoname and v not in aux:
+                        _verboseprint(f'Warning: Expected {v} to be an event in "{name}": "{form}" on {self.shortname}')
```
Comment (captain5050) [inline scripts/create_perf_json.py:1248]: Probably this can go back to an assert again.

File: scripts/create_perf_json.py [hunk 4]
```diff
@@ -1201,12 +1238,14 @@ def save_form(name, group, form, desc, locate, scale_unit, threshold,
                              'num_packages', 'num_cores', 'SYSTEM_TSC_FREQ',
                              'filter_tid', 'TSC', 'cha', 'config1',
                              'source_count', 'slots', 'thresh', 'has_pmem',
-                             'num_dies', 'num_cpus_online']:
+                             'num_dies', 'num_cpus_online', 'PEBS']:
```
Comment (captain5050) [inline scripts/create_perf_json.py:1241]: Probably not needed now.

**Review by captain5050 — APPROVED** [review]
A bunch of nits, but in general LGTM.

File: scripts/create_perf_json.py [hunk 5]
```diff
@@ -671,7 +701,9 @@ def extract_tma_metrics(self, csvfile: TextIO, pmu_prefix: str,
             'SPR': ('SPR', 'ADL/RPL', 'TGL', 'RKL', 'ICX', 'ICL', 'CNL', 'CPX', 'CLX',
                     'KBLR/CFL/CML', 'SKX', 'SKL/KBL', 'BDX', 'BDW', 'HSX', 'HSW', 'IVT',
                     'IVB', 'JKT/SNB-EP', 'SNB'),
-            "GRT": ("GRT"),
+            "GRT": ("GRT",),
```
Comment (edwarddavidbaker) [inline scripts/create_perf_json.py:704]: Remove extra comma inside parenthesis.

File: scripts/create_perf_json.py [hunk 6]
```diff
@@ -671,7 +701,9 @@ def extract_tma_metrics(self, csvfile: TextIO, pmu_prefix: str,
...
+            "MTL": ('MTL', 'ADL/RPL', 'TGL', 'RKL', 'ICL', 'CNL', 'KBLR/CFL/CML',
+                        'SKL/KBL', 'BDW', 'HSW', 'IVB', 'SNB'),
```
Comment (edwarddavidbaker) [inline scripts/create_perf_json.py:706]: * Nit: Line up 'SKL/KBL' with 'MTL' to match the rest of formatting.
* The platform hierarchy here is up-to-date outside of CPX and SPR-HBM. As you've included here, I also noticed that the KBLR/CFL includes CML. Should we update line 689 to also include `CML` to match `inheritance_rules` in https://github.com/intel/perfmon/blob/main/TMA_Metrics-full.xlsx ?

**Review by edwarddavidbaker — CHANGES_REQUESTED** [review]
Thanks for submitting Weilin!

**Comment by weilinwa** [comment]
@captain5050 and @edwarddavidbaker, I've updated the code based on your comments. Please take a look. Thanks!

File: scripts/create_perf_json.py (hunk 4 above)
Comment (weilinwa) [reply scripts/create_perf_json.py:1241]: This check seems to be executed before function `extract_pebs_formula(formula)` is called. Therefore, we still need to add this term here to continue the execution.

File: scripts/create_perf_json.py (hunk 6 above)
Comment (edwarddavidbaker) [reply scripts/create_perf_json.py]: Thanks Weilin.

File: scripts/create_perf_json.py (hunk 5 above)
Comment (edwarddavidbaker) [reply scripts/create_perf_json.py]: Thanks Weilin.

File: scripts/create_perf_json.py [hunk 7]
```diff
@@ -651,15 +651,17 @@ def extract_pebs_formula(formula):
...
                 new_formula = re.sub(m.re, new_string, new_formula, count=1)
 
         formula_list = re.split(OPS, new_formula)
         for element in formula_list:
             element = element.strip()
             if '$PEBS' in element:
                 event_name = element.split('*')[0]
-                new_element = "( " + event_name + ' * ' + event_name + ':retire_latency )'
+                mod = 'R' if '@' in event_name else ':R'
+                new_element = "( " + event_name + ' * ' + event_name + mod + ' )'
```
Comment (captain5050) [inline scripts/create_perf_json.py:664]: nit: maybe use an f-string for consistency with elsewhere.

**Review by captain5050 — APPROVED** [review]

**Comment by edwarddavidbaker** [comment]
@weilinwa MTL uncore updates in https://github.com/intel/perfmon/pull/147 fix the following error.

```
AssertionError: Expected UNC_HAC_ARB_COH_TRK_REQUESTS.ALL to be an event in "tma_info_system_dram_bw_use": "64 * (UNC_HAC_ARB_TRK_REQUESTS.ALL + UNC_HAC_ARB_COH_TRK_REQUESTS.ALL) / 1000000000 / duration_time" on MTL
```

**Review by edwarddavidbaker — APPROVED** [review]

**Comment by weilinwa** [comment]
> @weilinwa MTL uncore updates in #147 fix the following error.
> 
> ```
> AssertionError: Expected UNC_HAC_ARB_COH_TRK_REQUESTS.ALL to be an event in "tma_info_system_dram_bw_use": "64 * (UNC_HAC_ARB_TRK_REQUESTS.ALL + UNC_HAC_ARB_COH_TRK_REQUESTS.ALL) / 1000000000 / duration_time" on MTL
> ```

Thanks @edwarddavidbaker! I've rebased the branch.

**Comment by edwarddavidbaker** [comment]
@weilinwa Are Ian's conversations resolved in the latest series? If so please comment on them and mark as resolved.

File: scripts/create_perf_json.py (hunk 1 above)
Comment (weilinwa) [reply scripts/create_perf_json.py:672]: Yes, most of them could be matched by the min/max pattern. The remaining one(s) are replaced in this way.

**Comment by weilinwa** [comment]
> @weilinwa Are Ian's conversations resolved in the latest series? If so please comment on them and mark as resolved.

@edwarddavidbaker I just marked them as resolved.
